lib/relay-git.js

```diff
@@ -0,0 +1,115 @@
+const request = require('request');
+const { v4: uuid } = require('uuid');
+const tryJSONParse = require('./try-json-parse');
+const logger = require('./log');
+
+const GIT_CLIENT_REDIRECT_PATH = '/snykgit';
+
+module.exports = {
+  GIT_CLIENT_REDIRECT_PATH,
+  // no requestHandler here because we don't need any special
+  // HTTP-to-websocket transformation for git client requests
+  response: responseHandler,
+};
+
+// 1. Git request coming in over websocket conn (logged)
+// 2. Relay over HTTP conn to git client (logged)
+// 3. Get response over HTTP conn (logged)
+// 4. Send response over websocket conn
+function responseHandler(config, io) {
+  return (brokerToken) => (
+    { url, headers = {}, method, body = null, streamingID = '' } = {},
+    emit,
+  ) => {
+    // removing GIT_CLIENT_REDIRECT_PATH from request url;
+    if (url && url.startsWith(GIT_CLIENT_REDIRECT_PATH)) url = url.slice(GIT_CLIENT_REDIRECT_PATH.length);
+    const logContext = {
+      url,
+      method,
+      headers,
+      requestId: headers['snyk-request-id'] || uuid(),
+      streamingID,
+    };
+
+    logger.debug(logContext, 'received request over websocket connection');
+
+    const { GIT_CLIENT_URL, GIT_URL, GIT_USERNAME, GIT_PASSWORD } = config;
+    let errorStatus;
+    if (brokerToken) errorStatus = 'NO_GIT_CLIENT_ON_BROKER_SERVER';
+    else if (!GIT_CLIENT_URL) errorStatus = 'MISSING_GIT_CLIENT_URL';
+    else if (!GIT_URL || !GIT_USERNAME || !GIT_PASSWORD) errorStatus = 'MISSING_GIT_SERVER_CONFIG';
+    if (errorStatus) {
+      logger.error(
+        { ...logContext, errorStatus },
+        'received git client request with wronng configuration'
+      );
+      return emit({
+        status: 500,
```

Comment:
optional: bad request would be better?

Reply:
It's not a bad request, but a wrong internal configuration. The request might be fine, but the data to inject is not there.